safir/src/safir/metrics/_event_manager.py

```diff
@@ -176,6 +184,34 @@ async def publish(self, payload: P) -> EventMetadata:
         return event
 
 
+class MockEventPublisher(NoopEventPublisher, Generic[P]):
+    """Event publisher that quietly does nothing and records all payloads.
+
+    This is meant to be used in unit tests to enable assertions on published
+    payloads. It should NOT be used in any deployed application instances
+    because memory usage will grow unbounded with each published event.
+    """
+
+    def __init__(
+        self,
+        application: str,
+        event_class: type[AvroBaseModel],
```

Comment:
```suggestion
        event_class: type[P],
```

Reply:
`P` is the payload type here, whereas `event_class` is the dynamically smooshed payload and metadata type.

Reply:
Then the prototype for `EventPublisher` is also wrong, but I'm not sure that's correct. See the prototype for `construct_event`; we're using `P` as the type for the payload.

We don't have a separate type for the smooshed payload and metadata type. It would be ideal if we did, but since it's dynamically constructed, mypy cannot represent that. However, it is a child class of `P`, so using `P` for it is not incorrect, and is more accurate (and more constrained) than saying it's any `AvroBaseModel`.

Reply:
TLDR: Blech, definitely agreed that its inconsistent. What do you think about using a type alias to refer to the enriched event, like [this](https://github.com/lsst-sqre/safir/pull/348/commits/06ce821c1bc00d5f7673d142e81572e222a2b702)?

I don't love using the `P` parameter to mean both (even though I've already done it :smile: ) because I think it makes the code confusing and obscures the already-tricky difference between the two types. I think that's why I originally (and inconsistently :disappointed: ) declared the enriched type as `AvroBaseModel`, to differentiate it.

As a side note, I don't understand why mypy thinks it's OK that [`construct_event` returns `P`](https://github.com/lsst-sqre/safir/blob/5e1c4f5b6798f99a8a66ed70afea487c885fc609/safir/src/safir/metrics/_event_manager.py#L61-L84) and [`publish` returns the `construct_event` return value but says it's an `EventMetadata`](https://github.com/lsst-sqre/safir/blob/5e1c4f5b6798f99a8a66ed70afea487c885fc609/safir/src/safir/metrics/_event_manager.py#L151-L154), when [`P` is bound to `EventPayload`](https://github.com/lsst-sqre/safir/blob/5e1c4f5b6798f99a8a66ed70afea487c885fc609/safir/src/safir/metrics/_event_manager.py#L28-L28)

...looks like mypy is saying that the return value from `construct_event` is `Any`:
```
async def publish(self, payload: P) -> EventMetadata:
        event = self.construct_event(payload)
        reveal_type(event)
        await self._manager.publish(event, self._publisher, self._schema_info)
        return event
```
```nox > mypy --namespace-packages --explicit-package-bases noxfile.py safir/src safir-arq/src safir-logging/src safir/tests
safir/src/safir/metrics/_event_manager.py:153: note: Revealed type is "Any"
```
It makes sense to me that mypy doesn't know what the type is, but I don't get why mypy  lets us call it something else and then just starts throwing `Any` around.  I'm glad it does in this case, otherwise this wouldn't work :smile: :shrug: 

Fwiw, The `EnrichedEvent` change doesn't help here, it still subs `Any` for the `P`:
```nox > mypy --namespace-packages --explicit-package-bases noxfile.py safir/src safir-arq/src safir-logging/src safir/tests
safir/src/safir/metrics/_event_manager.py:167: note: Revealed type is "Union[Any, safir.metrics._models.EventMetadata]"```

Reply:
I'm not a fan of the union because I think that will make the typing more tedious, not less. If someone wants to see the enriched fields that we add to the type in the return value from publish, now they'll have to add a cast or an `isinstance` check or mypy will complain that those fields are missing from the `P` branch of the union.

The root problem here is that the semantics you want for the return value from `publish` are intersection, not union. The returned type has to be a child type of *both* `P` and `EventMetadata`, and is therefore a valid instance of both of them. Unfortunately, there isn't a simple way to represent this because `TypeVar` only takes a single bound.

Because of that, I feel like the proposed change is even more inaccurate in the typing, since it semantically says that this model can be either some instance of `EventPayload` *or* `EventMetadata`, but really it has to be both, and the union also loses the strong association between the constructor parameter type and the input type for publish (you could pass a simple instance of `EventMetadata` to `publish` now, which is wrong).

I think the best we can do here is to parameterize `EventPublisher` with two `TypeVar` parameters, one of which has a bound of `EventPayload` and represents the application-side event model and the other of which has a bound of `EventMetadata` and represents the enriched event. Call them `P` and `E`. Then, `__init__` takes a parameter of type `type[E]`, `construct_event` takes `P` and returns `E`, and `publish` takes `P` and returns `E`.
I think that's as good as we can get in representing the type system given the lack of multiple bounds, the lack of a way to define one generic type parameter is a subclass of another, and the fact mypy is always going to bind `E` to `Any` right now anyway.

But the problem with doing *that* is that when you construct the `EventPublisher` class, you have to provide `P` since it cannot be determined from the constructor arguments, which I think means you have to provide `E` as well, and there's no valid value of `E` that you can provide. You end up having to write an explicit `Any`, I think.

This is why I just used `P` for everything when I suggested this type scheme originally. It seems strictly better in terms of type algebra than using `AvroBaseModel`, which is much less constrained. And it is actually *correct*; it's always valid to declare a function returning some class as returning any superclass of that class, and `P` is a superclass of every class involved, so the type system here where `P` is used everywhere is formally correct. It just discards information because it loses the fact that `EventMetadata` has been mixed in and those fields are therefore guaranteed to be present (which is why the return type of `publish`).

Reply:
I think the critical thing here is to make sure that the resulting type of the publisher object is `EventPublisher[SomeConcreteMetricsEvent]`, not `EventPublisher[Any]`. That does mean that I'm wrong and you can't omit the `[]` when creating the publisher. You do need to ensure that's bound to the input event model (*not* the enriched type).

Reply:
`P` works for me.

> But the problem with doing that is that when you construct the EventPublisher class, you have to provide P since it cannot be determined from the constructor arguments, which I think means you have to provide E as well, and there's no valid value of E that you can provide. You end up having to write an explicit Any, I think.

Yep, I tried this and this is more or less where I wound up.

> That does mean that I'm wrong and you can't omit the [] when creating the publisher. You do need to ensure that's bound to the input event model (not the enriched type).

Assuming you mean [this](https://github.com/lsst-sqre/safir/pull/348/commits/9efbb16f58ae84300ce664dc22190bd8605c3a89), it doesn't seem to make a difference, `mypy` can figure it out, I think because we cast the `build_publisher_for_model` arg [here](https://github.com/lsst-sqre/safir/blob/tickets%2FDM-47764%2Fmetrics-events-unittest-support/safir/src/safir/metrics/_event_manager.py#L323-L329).

But it doesn't seem to hurt anything to be explicit about it, and I quite obviously have a lot to learn about how `mypy` deals with generics, so I put them back in :)